File: motor/division.py
"""Auto-splitter (§5.3): divide un HTML completo en capítulos usando
cabeceras <h1>, <h2> y <h3> como puntos de corte."""

import re

from motor.limpieza import texto_plano
from motor.modelo import Chapter

RE_ENCABEZADO = re.compile(r'<h([1-3])(?:\s[^>]*)?>(.*?)</h\1>', re.DOTALL | re.IGNORECASE)


def dividir_en_capitulos(
    html: str, titulos: list[str] | None = None, start_num: int = 1
) -> list[Chapter]:
    """Convierte un HTML completo en capítulos. La cabecera de cada corte se
    extrae del cuerpo y se usa como título cuando no hay lista de títulos.
    El contenido anterior a la primera cabecera forma su propio capítulo."""
    matches = list(RE_ENCABEZADO.finditer(html))

    if not matches:
        titulo = titulos[0] if titulos else f'Capítulo {start_num}'
        return [Chapter(titulo=titulo, html_cuerpo=html.strip())]

    capitulos: list[Chapter] = []
    contador = start_num

    pre = html[:matches[0].start()].strip()
    if pre:
        titulo = titulos[0] if titulos else f'Capítulo {contador}'
        capitulos.append(Chapter(titulo=titulo, html_cuerpo=pre))
        contador += 1

    limites = [m.start() for m in matches] + [len(html)]
    for i, match in enumerate(matches):
        cuerpo = RE_ENCABEZADO.sub('', html[limites[i]:limites[i + 1]], count=1).strip()
        texto = texto_plano(match.group(2))
        indice = contador - start_num
        titulo = titulos[indice] if titulos and indice < len(titulos) else (texto or f'Capítulo {contador}')
        capitulos.append(Chapter(titulo=titulo, html_cuerpo=cuerpo))
        contador += 1

    return capitulos

File: app_flutter/lib/motor/division.dart
import 'modelo.dart';
import 'limpieza.dart' show textoPlano;

final _reEncabezado = RegExp(r'<h([1-3])(?:\s[^>]*)?>(.*?)</h\1>', dotAll: true, caseSensitive: false);
final _reEncabezadoVacio = RegExp(r'<h[1-3](?:\s[^>]*)?>(?:\s|&nbsp;)*</h[1-3]>', caseSensitive: false);

List<Chapter> dividirEnCapitulos(String html, {List<String>? titulos, int startNum = 1}) {
  final matches = _reEncabezado.allMatches(html).where((m) => textoPlano(m.group(2)!).isNotEmpty).toList();

  if (matches.isEmpty) {
    String titulo = (titulos != null && titulos.isNotEmpty) ? titulos[0] : 'Capítulo $startNum';
    return [Chapter(titulo: titulo, htmlCuerpo: html.replaceAll(_reEncabezadoVacio, '').trim())];
  }

  List<Chapter> capitulos = [];
  int contador = startNum;

  String pre = html.substring(0, matches[0].start).replaceAll(_reEncabezadoVacio, '').trim();
  if (pre.isNotEmpty) {
    String titulo = (titulos != null && titulos.isNotEmpty) ? titulos[0] : 'Capítulo $contador';
    capitulos.add(Chapter(titulo: titulo, htmlCuerpo: pre));
    contador++;
  }

  List<int> limites = matches.map((m) => m.start).toList()..add(html.length);
  for (int i = 0; i < matches.length; i++) {
    String subHtml = html.substring(limites[i], limites[i+1]);
    String cuerpo = subHtml.replaceFirst(_reEncabezado, '');
    cuerpo = cuerpo.replaceAll(_reEncabezadoVacio, '').trim();
    String texto = textoPlano(matches[i].group(2)!);
    int indice = contador - startNum;
    String titulo = (titulos != null && indice < titulos.length) ? titulos[indice] : (texto.isNotEmpty ? texto : 'Capítulo $contador');
    capitulos.add(Chapter(titulo: titulo, htmlCuerpo: cuerpo));
    contador++;
  }

  return capitulos;
}
